### ISSUE TITLE
Generating parser from grammar doesn't work on Windows with shell other than `cmd`

### ISSUE THREAD

**Issue by binyomen** [nvim-treesitter/nvim-treesitter#3316]
### Describe the bug

The Win32 section of `make_directory_change_for_command` uses syntax that assumes the current shell is `cmd`. If the current shell isn't `cmd`, then it may behave differently. In my case with the shell set to `pwsh`, trying to install the gitignore parser prints out "Generating source files from grammar.js..." and then nothing else. The `tree-sitter` CLI no longer seems to be running, and looking in the cache directory the parser files have not been generated.

### To Reproduce

You need to be on Windows.

Set your shell to `pwsh` with:

```lua
    vim.opt.shell = 'pwsh'
    vim.opt.shellcmdflag = '-NoLogo -NoProfile -ExecutionPolicy RemoteSigned -Command [Console]::InputEncoding=[Console]::OutputEncoding=[System.Text.Encoding]::UTF8;'
    vim.opt.shellredir = '2>&1 | Out-File -Encoding UTF8 %s; exit $LastExitCode'
    vim.opt.shellpipe = '2>&1 | Out-File -Encoding UTF8 %s; exit $LastExitCode'
    vim.opt.shellquote = ''
    vim.opt.shellxquote = ''

    vim.opt.shelltemp = false
```

Also make sure `gitignore` is included in your `ensure_installed` list under `require('nvim-treesitter.configs').setup`.

When starting `nvim`, "Generating source files from grammar.js..." will be printed but the installation process won't continue past there.

### Expected behavior

The gitignore parser to be successfully installed.

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()
========================================================================
## Installation
  - OK: `tree-sitter` found 0.20.6 (ccd6bf554d922596ce905730d98a77af368bba5c) (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v18.7.0 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `clang` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
    Version: clang version 14.0.6
  - OK: Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I J
  - html           ✓ ✓ ✓ ✓ ✓
  - json           ✓ ✓ ✓ ✓ .
  - gitignore      ✓ . . . .
  - markdown       ✓ . ✓ . ✓
  - python         ✓ ✓ ✓ ✓ ✓
  - cpp            ✓ ✓ ✓ ✓ ✓
  - fish           ✓ ✓ ✓ ✓ ✓
  - javascript     ✓ ✓ ✓ ✓ ✓
  - lua            ✓ ✓ ✓ ✓ ✓
  - css            ✓ . ✓ ✓ ✓
  - yaml           ✓ ✓ ✓ ✓ ✓
  - vim            ✓ ✓ ✓ . ✓
  - toml           ✓ ✓ ✓ ✓ ✓
  - markdown_inline✓ . . . ✓
  - c              ✓ ✓ ✓ ✓ ✓
  - query          ✓ ✓ ✓ ✓ ✓
  - comment        ✓ . . . .
  - regex          ✓ . . . .
  - typescript     ✓ ✓ ✓ ✓ ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}
```

### Output of `nvim --version`

```text
NVIM v0.7.2
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compiled by runneradmin@fv-az276-503

Features: -acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM\sysinit.vim"
  fall-back for $VIM: "C:/Program Files/nvim/share/nvim"

Run :checkhealth for more info
```

### Additional context

I can think of a couple possible ways to address this in `make_directory_change_for_command`.

One is to make the use of `cmd` explicit. Something like this maybe:

```lua
return string.format('cmd /c "pushd %s & %s & popd"', vim.fn.shellescape(dir), command)
```

It's possible there are shells on Windows ([nushell](https://github.com/nushell/nushell) maybe?) where this kind of escaping doesn't work. Perhaps a more robust solution would be to set the shell settings to default on Windows before running any commands. You can do this with something like:

```lua
vim.opt.shell = 'cmd.exe'
vim.opt.shellcmdflag = '/s /c'
vim.opt.shellredir = '>%s 2>&1'
vim.opt.shellpipe = '>%s 2>&1'
vim.opt.shellquote = ''
vim.opt.shellxquote = '"'

vim.opt.shelltemp = true
```

I'd be curious to hear your opinions on the best approach here.

**Comment by binyomen** [comment]
Another possible solution would be to pass a list rather than a string to `vim.fn.system`, which won't execute the command under a shell. If that's possible, it could avoid some of the shell option weirdness.

**Comment by yehorb** [comment]
I encountered the issue too. It looks like `make_directory_change_for_command` is not the root problem.  `make_directory_change_for_command` is only used in `*Sync` commands - `:TSUpdateSync,:TSInstallSync` etc. Changing `make_directory_change_for_command` to check for PowerShell and providing the correct change directory command (https://github.com/yehorb/nvim-treesitter/commit/d00e5db84274d70af173110d4e2e736cc4f58a4c) allowed me to install `gitignore` parser via `:TSInstallSync gitignore`, but it does not solve the async install issue. It is still stuck on `Generating source files from grammar.js...`

**Comment by yehorb** [comment]
Looks like the issue is somehow with the `.ps1` shim generated by `npm`. If I force the `.cmd` shim in the code (`cmd = vim.fn.exepath("tree-sitter") .. ".cmd"`) it works in non-`Sync` way. Looks like `luv.spawn` does not know what to do with `.ps1` shim. And if you have PowerShell as your `vim.o.shell`, then the `.ps1` shim will be called.
To fix the issue right now, you can delete the `tree-sitter.ps1` from your `npm bin` directory.

**Comment by yehorb** [comment]
The [`CreateProcessW `](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessw) that is used by `libuv` to spawn the new process was never intended to call batch files, docs say so right at the beginning. Why the `.cmd` shims "just work ™" is unknown to me. Making it work requires some serious changes (something along the lines of https://github.com/nvim-treesitter/nvim-treesitter/commit/c9e0c87f1bba18502cdf2b4588fcb49239327329), not sure what is the best fix.

**Comment by akriese** [comment]
I just encountered the same (or at least similar issue), where tree-sitter would "install" all parsers at every start of neovim. Figured out the problem to be the `&` (background process in pwsh, might lead to out-of-order-execution) instead `&&` (chain pipe in pwsh) in `make_directory_change_for_command`. As this seems to be the only case where the `&` syntax is used in the installation process, this should be easily fixed by doing something like:
```lua
function M.make_directory_change_for_command(dir, command)
  if fn.has "win32" == 1 then
    local chain = vim.o.shell == "pwsh" and "&&" or "&"
    return string.format("pushd %s %s %s %s popd", dir, chain, command, chain)
  else
    return string.format("cd %s;\n %s", dir, command)
  end
end
```

This fixed my problem, that tree-sitter removed the downloaded repo before everything was compiled. If this is worth an individual issue, I can create one. But I thought I'd add my thoughts here, as it is very related :)
